gufe/chemicalstate.py

```diff
@@ -1,6 +1,98 @@
 # This code is part of OpenFE and is licensed under the MIT license.
 # For details, see https://github.com/OpenFreeEnergy/gufe
 
+from typing import Dict, Optional
+
+import numpy as np
+
+from .component import Component
+
+
 class ChemicalState:
```

Comment:
This class represents a particular composition of components, so names like `ChemicalSystem`, `ComponentCollection`, or `ChemicalComponents` might be more appropriate. This is not a strong recommendation, however.

It's worth noting that this is very similar to the [`openff.toolkit.topology.Topology`](https://open-forcefield-toolkit.readthedocs.io/en/0.10.3/api/generated/openff.toolkit.topology.Topology.html#openff.toolkit.topology.Topology) class, which is meant to hold multiple chemical components whose chemical identities, positions, and potentially periodic cell are uniquely specified. Essentially all of the information you seek to store is already represented in a much more robust manner in `Topology`, and if the path to upgrade to `Topology` is desired, it's worth building in consistency between a current minimal interface and the `Topology` API. The one thing that is missing is labels/tags/identifiers for each component, but it's worth noting the `Molecule.name` field is already available to name the components.

I'd check with @j-wags about what expected changes are coming now that biopolymer support is being added. I don't imagine the `Molecule` object will experience significant API-breaking changes, however.

Reply:
@jchodera I believe hashability, and therefore immutability, where the major concerns with using `openff.toolkit.topology.Topology` directly as the `ChemicalState`, since our aim is to use the `ChemicalState` *itself* as a node in a `networkx`-based `AlchemicalNetwork`.

I think we recognize that both `Topology` and `ChemicalState` perform similar roles, but they serve different contexts: `Topology` is designed for the application of forcefield parameters on a system, while `ChemicalState` is designed for the hashability needs of the graph representation in an `AlchemicalNetwork`. For many protocols we anticipate generating `Topology`s from these hashable representations, but this sits downstream in our use of `ChemicalState`.